--- linux-6.6.51/drivers/net/phy/phylink.c	2024-12-05 13:07:50.568471200 +0900
+++ linux4microchip-fpga-2024.09/drivers/net/phy/phylink.c	2024-12-04 16:41:30.883341100 +0900
@@ -2726,6 +2726,16 @@
 }
 EXPORT_SYMBOL_GPL(phylink_ethtool_set_eee);
 
+/**
+ * phylink_init_phydev() - initialize phydev associated to phylink
+ * @pl: a pointer to a &struct phylink returned from phylink_create()
+ */
+int phylink_init_phydev(struct phylink *pl)
+{
+	return phy_init_hw(pl->phydev);
+}
+EXPORT_SYMBOL_GPL(phylink_init_phydev);
+
 /* This emulates MII registers for a fixed-mode phy operating as per the
  * passed in state. "aneg" defines if we report negotiation is possible.
  *
--- linux-6.6.51/include/linux/phylink.h	2024-12-05 13:07:57.485964300 +0900
+++ linux4microchip-fpga-2024.09/include/linux/phylink.h	2024-12-04 16:41:37.803825900 +0900
@@ -689,6 +689,7 @@
 int phylink_mii_ioctl(struct phylink *, struct ifreq *, int);
 int phylink_speed_down(struct phylink *pl, bool sync);
 int phylink_speed_up(struct phylink *pl);
+int phylink_init_phydev(struct phylink *pl);
 
 #define phylink_zero(bm) \
 	bitmap_zero(bm, __ETHTOOL_LINK_MODE_MASK_NBITS)
