--- linux-6.6.51/arch/riscv/Kconfig.socs	2024-12-05 13:07:38.105306700 +0900
+++ linux4microchip-fpga-2024.09/arch/riscv/Kconfig.socs	2024-12-04 16:41:17.450018400 +0900
@@ -5,6 +5,8 @@
 
 config SOC_MICROCHIP_POLARFIRE
 	bool "Microchip PolarFire SoCs"
+	select DMA_GLOBAL_POOL
+	select RISCV_NONSTANDARD_CACHE_OPS
 	help
 	  This enables support for Microchip PolarFire SoC platforms.
 
--- linux-6.6.51/arch/riscv/include/asm/vendorid_list.h	2024-12-05 13:07:38.188085500 +0900
+++ linux4microchip-fpga-2024.09/arch/riscv/include/asm/vendorid_list.h	2024-12-04 16:41:17.527810300 +0900
@@ -6,6 +6,7 @@
 #define ASM_VENDOR_LIST_H
 
 #define ANDESTECH_VENDOR_ID	0x31e
+#define MICROCHIP_VENDOR_ID	0x029
 #define SIFIVE_VENDOR_ID	0x489
 #define THEAD_VENDOR_ID		0x5b7
 
--- linux-6.6.51/arch/riscv/kernel/smp.c	2024-12-05 13:07:38.221994500 +0900
+++ linux4microchip-fpga-2024.09/arch/riscv/kernel/smp.c	2024-12-04 16:41:17.560722200 +0900
@@ -39,6 +39,7 @@
 unsigned long __cpuid_to_hartid_map[NR_CPUS] __ro_after_init = {
 	[0 ... NR_CPUS-1] = INVALID_HARTID
 };
+EXPORT_SYMBOL_GPL(__cpuid_to_hartid_map);
 
 void __init smp_setup_processor_id(void)
 {
