--- linux-6.6.51/drivers/net/dsa/microchip/ksz_common.c	2024-12-05 13:07:49.030585800 +0900
+++ linux4microchip-fpga-2024.09/drivers/net/dsa/microchip/ksz_common.c	2024-12-04 16:41:29.239738200 +0900
@@ -511,8 +511,7 @@
 	regmap_reg_range(0x111a, 0x111d),
 	regmap_reg_range(0x1122, 0x1127),
 	regmap_reg_range(0x112a, 0x112b),
-	regmap_reg_range(0x1136, 0x1139),
-	regmap_reg_range(0x113e, 0x113f),
+	regmap_reg_range(0x1134, 0x113f),
 	regmap_reg_range(0x1400, 0x1401),
 	regmap_reg_range(0x1403, 0x1403),
 	regmap_reg_range(0x1410, 0x1417),
@@ -541,8 +540,7 @@
 	regmap_reg_range(0x211a, 0x211d),
 	regmap_reg_range(0x2122, 0x2127),
 	regmap_reg_range(0x212a, 0x212b),
-	regmap_reg_range(0x2136, 0x2139),
-	regmap_reg_range(0x213e, 0x213f),
+	regmap_reg_range(0x2134, 0x213f),
 	regmap_reg_range(0x2400, 0x2401),
 	regmap_reg_range(0x2403, 0x2403),
 	regmap_reg_range(0x2410, 0x2417),
